File: Project/C++/src/solver.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <boost/algorithm/string.hpp>
#include "transaction.cpp"
#include "rule.cpp"
#include "classification.cpp"
using namespace std;

class solver
{
private:
    vector<classification> classifications;
public:
    solver(string rulespath)
    {
        ifstream inputFile;
        inputFile.open(rulespath, ios::in);

        string line = "";
        vector<rule> rules;
        
        while (getline(inputFile, line))
        {
            vector<string> temp;
            boost::split(temp, line, [](char c) { return c == ';'; });
            rules.push_back(rule(temp));
        }

        sort(rules.begin(), rules.end(), compareRule);

        classification c = classification(rules[0].getClassification());
        c.addRule(rules[0]);
        classifications.push_back(c);

        for (int i = 1; i < rules.size(); i++)
        {
            if (rules[i].getClassification() != c.getClassification())
            {
                c = classification(rules[i].getClassification());
                c.addRule(rules[i]);
                classifications.push_back(c);
            }
            else
            {
                c.addRule(rules[i]);
            }
        }
        cout << "Completed reading rules..." << endl;
    }

    void solve(string in, string out)
    {

        vector<transaction> transactions;

        ifstream infile;
        infile.open(in, ios::in);

        string line;
        while (getline(infile, line))
        {
            vector<string> temp;
            boost::split(temp, line, [](char c) { return c == ';'; });
            transactions.push_back(transaction(temp));
        }

        ofstream outfile;
        outfile.open(out, ios::out);

        for (int i = 0; i < transactions.size(); i++)
        {
            string s = "";
            for (int j = 0; j < classifications.size(); j++)
            {
                if(classifications[j].matches(transactions[i])){
                    s.append(transactions[i].toString());
                    s.append(to_string(classifications[j].getClassification()));
                    s.append("\n");
                }
            }
            outfile << s;
        }

        infile.close();
        outfile.close();
    }
};

File: Project/Java/Solver.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Solver {

  static final String cvsSplitBy = ";";
  private List<Classification> classifications;

  public Solver(String filename) {
    String line = "";
    ArrayList<Rule> rules = new ArrayList<>(20000);
    classifications = new ArrayList<>(50);

    try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
      while ((line = br.readLine()) != null) {
        rules.add(new Rule(line.split(cvsSplitBy)));
      }
    } catch (FileNotFoundException e) {
      e.printStackTrace();
    } catch (IOException e) {
      e.printStackTrace();
    }

    Collections.sort(rules);

    writeToFile(rules);

    Classification c = new Classification(rules.get(0).getClassification());
    c.addRule(rules.get(0));
    classifications.add(c);

    for (int i = 1; i < rules.size(); i++) {
      if (rules.get(i).getClassification() != c.getClassification()) {
        c = new Classification(rules.get(i).getClassification());
        c.addRule(rules.get(i));
        classifications.add(c);
      } else {
        c.addRule(rules.get(i));
      }
    }
    rules = null;
    System.out.println("Rules loaded...");
  }

  public void writeToFile(List<Rule> rules){
    File out = new File("teste.txt");
    try (PrintWriter pw = new PrintWriter(out)){
      for(Rule r : rules){
        pw.println(r.toString());
      }
    } catch (Exception e) {
      //TODO: handle exception
    }
  }

  public void solve(String input, String output) {
    String line;
    ArrayList<Transaction> transactions = new ArrayList<>(10000);

    try (BufferedReader br = new BufferedReader(new FileReader(input))) {
      while ((line = br.readLine()) != null) {
        transactions.add(new Transaction(line.split(cvsSplitBy)));
      }
    } catch (FileNotFoundException e) {
      e.printStackTrace();
    } catch (IOException e) {
      e.printStackTrace();
    }

    File csvOutputFile = new File(output);

    long start = System.currentTimeMillis();
    try (PrintWriter pw = new PrintWriter(csvOutputFile)) {
      for (Transaction t : transactions) {
        for (Classification cl : classifications) {
          if (cl.matches(t)) {
            pw.println(t.toString() + cl.getClassification());
          }
        }
      }
      double timeTaken = (System.currentTimeMillis() - start) / 1000.0;
      System.out.println(transactions.size() / timeTaken + " Transacions per second");
    } catch (FileNotFoundException e) {
      e.printStackTrace();
    }
  }
}
